### ISSUE TITLE
static initializers create LRUCache too many times at process start

### ISSUE THREAD

**Issue by mdcallag** [facebook/rocksdb#3176]
Reproduction is: run mysqld with gdb after putting breakpoint on the LRUCache ctor and count how many times that is run from the statistic initializers in options_helper.h. It comes from this code and the template above it.
https://github.com/facebook/rocksdb/blob/master/options/options_helper.h#L391

There are ~10 calls from the sequence:
options_helper.h -> ColumnFamilyOptions -> BlockBasedTableFactory -> NewLRUCache -> LRUCache

Then there is one more from the Rdb_cf_options ctor. That is MyRocks code which suffers from what ColumnFamilyOptions does.
```
#11 0x0000000000cf52fe in rocksdb::ColumnFamilyOptions::ColumnFamilyOptions (this=0x2605918) at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options.cc:99
#12 0x0000000000b922ea in myrocks::rocksdb_init_func (p=<optimized out>) at /data/users/mcallaghan/mysql.me/5.6/storage/rocksdb/././rdb_cf_options.h:47
#13 0x0000000000778778 in ha_initialize_handlerton (plugin=0x3afad78) at /data/users/mcallaghan/mysql.me/5.6/sql/handler.cc:655
#14 0x0000000000901e7b in plugin_initialize (plugin=plugin@entry=0x3afad78) at /data/users/mcallaghan/mysql.me/5.6/sql/sql_plugin.cc:1163
#15 0x0000000000902b28 in plugin_init (argc=argc@entry=0x1bf5368 <remaining_argc>, argv=0x265f1e0, flags=flags@entry=0)
    at /data/users/mcallaghan/mysql.me/5.6/sql/sql_plugin.cc:1465
#16 0x0000000000768375 in init_server_components () at /data/users/mcallaghan/mysql.me/5.6/sql/mysqld.cc:5851
#17 0x000000000076ba25 in mysqld_main (argc=127, argv=0x265f1e0) at /data/users/mcallaghan/mysql.me/5.6/sql/mysqld.cc:7198
#18 0x00007ffff59ac858 in __libc_start_main () from /usr/local/fbcode/gcc-5-glibc-2.23/lib/libc.so.6
```

Then there is the real one that MyRocks will use
```
#18 myrocks::Rdb_cf_options::init (this=0x26058c0, table_options=..., prop_coll_factory=...,
    default_cf_options=0x263fc95 "write_buffer_size=64m;target_file_size_base=32m;max_bytes_for_level_base=512m;level0_file_num_compaction_trigger=4;level0_slowdown_writes_trigger=20;level0_stop_writes_trigger=30;max_write_buffer_numb"..., override_cf_options=0x12d4c5f "")
    at /data/users/mcallaghan/mysql.me/5.6/storage/rocksdb/rdb_cf_options.cc:56
#19 0x0000000000b9234b in myrocks::rocksdb_init_func (p=<optimized out>) at /data/users/mcallaghan/mysql.me/5.6/storage/rocksdb/ha_rocksdb.cc:4152
#20 0x0000000000778778 in ha_initialize_handlerton (plugin=0x3afad78) at /data/users/mcallaghan/mysql.me/5.6/sql/handler.cc:655
#21 0x0000000000901e7b in plugin_initialize (plugin=plugin@entry=0x3afad78) at /data/users/mcallaghan/mysql.me/5.6/sql/sql_plugin.cc:1163
#22 0x0000000000902b28 in plugin_init (argc=argc@entry=0x1bf5368 <remaining_argc>, argv=0x265f1e0, flags=flags@entry=0)
    at /data/users/mcallaghan/mysql.me/5.6/sql/sql_plugin.cc:1465
#23 0x0000000000768375 in init_server_components () at /data/users/mcallaghan/mysql.me/5.6/sql/mysqld.cc:5851
#24 0x000000000076ba25 in mysqld_main (argc=127, argv=0x265f1e0) at /data/users/mcallaghan/mysql.me/5.6/sql/mysqld.cc:7198
#25 0x00007ffff59ac858 in __libc_start_main () from /usr/local/fbcode/gcc-5-glibc-2.23/lib/libc.so.6
#26 0x00000000007343e9 in _start () at ../sysdeps/x86_64/start.S:118
```

An example with all of the stack traces is here - https://gist.github.com/mdcallag/80f9e4a51561124442254951d985ee9b

**Comment by maysamyabandeh** [comment]
There are two cases:
1) MyRocks should not have invoked the CF initialization that many times.
2) RocksDB should not re-create LRU cache that many time.
The latter can be avoid by myrocks setting the options properly as the option sanitizer in rocksdb sets the LRU automatically when the user failed to do so:
```
 41   if (table_options_.no_block_cache) {
 42     table_options_.block_cache.reset();
 43   } else if (table_options_.block_cache == nullptr) {
 44     table_options_.block_cache = NewLRUCache(8 << 20);
 45   }
```

**Comment by mdcallag** [comment]
With respect to 1. above, 90% of this comes from static initializers. MyRocks isn't invoking it. It comes from
```
#11 0x0000000000cf52fe in rocksdb::ColumnFamilyOptions::ColumnFamilyOptions (
    this=0x1c459a0 <int rocksdb::offset_of<std::shared_ptr<rocksdb::MergeOperator>, rocksdb::ColumnFamilyOptions>(std::shared_ptr<rocksdb::MergeOperator> rocksdb::ColumnFamilyOptions::*)::obj>) at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options.cc:99
#12 0x000000000067368a in rocksdb::offset_of<std::shared_ptr<rocksdb::MergeOperator>, rocksdb::ColumnFamilyOptions> (member=&rocksdb::ColumnFamilyOptions::merge_operator)
    at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:382
#13 __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:583
#14 0x000000000117d9ee in __libc_csu_init (argc=9, argv=0x7fffffffe318, envp=0x7fffffffe368) at elf-init.c:88
#15 0x00007ffff59ac7df in __libc_start_main () from /usr/local/fbcode/gcc-5-glibc-2.23/lib/libc.so.6
#16 0x00000000007343e9 in _start () at ../sysdeps/x86_64/start.S:118
```

**Comment by maysamyabandeh** [comment]
Thanks @mdcallag. It seems to be known issue with local static variables in inline functions. Let me give it a try to moving it to .cc and see if it handles the issue: https://stackoverflow.com/questions/11962918/local-static-variable-is-instantiated-multiple-times-why
```
386 inline int offset_of(T1 T2::*member) {
387   static T2 obj;
388   return int(size_t(&(obj.*member)) - size_t(&obj));
389 }
```

**Comment by maysamyabandeh** [comment]
The problem is that each .cc file will have a separate copy of this static variable and that is why its constructor is invoked multiple times. Since this is a temporary variable used only to get the offset of struct fields, it does not occur to me as a bug. The overhead of having 10 of this object is also seems to be negligible. It would still good to avoid it though.
@mdcallag let me know if the above understanding is inaccurate.

**Comment by mdcallag** [comment]
Maybe not a bug. But not elegant and an accident waiting to happen in the future when more code gets associated with that call stack.

**Comment by mdcallag** [comment]
Actually, I will modify the opinion. The number of times the ctor/dtor for LRUCache gets called is a function of the number of entries in cf_options_type_info that satisfy some condition.
https://github.com/facebook/rocksdb/blob/master/options/options_helper.h#L391

You can see that from the line numbers here:
https://gist.github.com/mdcallag/80f9e4a51561124442254951d985ee9b

This isn't a once per file thing.

**Comment by mdcallag** [comment]
I still disagree. Using data from the trace I shared above -- https://gist.github.com/mdcallag/80f9e4a51561124442254951d985ee9b

I grep'd for "options_helper" and then removed all lines with options_helper.h:382 and what remains is below. These are from entries in that static data structure that each trigger create/delete of LRUCache. This occurs once per each of those entries, not once per file.

```
#13 __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:406
#13 __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:451
    at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:504
    at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:545
#13 __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:556
#13 __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:559
#13 __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:572
#13 __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:575
#13 __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:579
#13 __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at /data/users/mcallaghan/mysql.me/5.6/rocksdb/options/options_helper.h:583
```

**Comment by maysamyabandeh** [comment]
The number of lines in your paste is 10. The number of cc files that include options_helper.h are also 10.
```
~/rocksdb]$ find . -name "*.cc" -exec grep options_helper.h /dev/null {} \; | grep -v test.cc | grep -v db_stress | wc -l
10
```
My assumption was that what trigger the static initialization is different from file to file and that is why we see different line numbers in the stack traces. I could be wrong though. Given that the fix is ready in https://github.com/facebook/rocksdb/pull/3178 it should be easy to verify. When I tried it on mysqld I found only one initialization of ColumnFamilyOptions.
